plugin/evm/message/handshake/upgrade_config_message.go

```diff
@@ -0,0 +1,117 @@
+// (c) 2019-2020, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package handshake
+
+import (
+	"github.com/ava-labs/subnet-evm/params"
+	"github.com/ava-labs/subnet-evm/precompile/modules"
+	"github.com/ethereum/go-ethereum/crypto"
+)
+
+type rawPrecompileUpgrade struct {
+	Key   string `serialize:"true"`
+	Bytes []byte `serialize:"true"`
+}
+
+type networkUpgradeConfigMessage struct {
+	OptionalNetworkUpgrades *params.OptionalNetworkUpgrades `serialize:"true,nullable"`
+	// Config for modifying state as a network upgrade.
+	StateUpgrades []params.StateUpgrade `serialize:"true"`
+	// Config for enabling and disabling precompiles as network upgrades.
+	PrecompileUpgrades []rawPrecompileUpgrade `serialize:"true"`
+}
+
+type UpgradeConfigMessage struct {
+	bytes []byte
+	hash  []byte
+}
+
+func (u *UpgradeConfigMessage) Bytes() []byte {
+	return u.bytes
+}
+
+func (u *UpgradeConfigMessage) ID() []byte {
+	return u.hash
+}
+
+// Attempts to parse a networkUpgradeConfigMessage from a []byte
+//
+// This function attempts to parse a stream of bytes as a
+// networkUpgradeConfigMessage (as serialized from
+// UpgradeConfigToNetworkMessage).
+//
+// The function returns a reference of *params.UpgradeConfig
+func NewUpgradeConfigMessageFromBytes(bytes []byte) (*params.UpgradeConfig, error) {
```

Comment:
why is this returning `*params.UpgradeConfig` instead of `UpgradeConfigMessage`?

Reply:
Can we rename the function to reflect that?

Reply:
Is the name better now @ceyonur ?

Reply:
@ceyonur I think now it is better. I dropped all the intermediary types and I added [MarshalBinary](https://github.com/ava-labs/subnet-evm/blob/handshake-codec/params/config.go#L197-L244), [UnmarshalBinary](https://github.com/ava-labs/subnet-evm/blob/handshake-codec/params/config.go#L246-L296) and [Hash](https://github.com/ava-labs/subnet-evm/blob/handshake-codec/params/config.go#L189-L195) to the `UpgradeConfig` struct. That way it is cleaner and much simpler.

I'm working now to re-structure the tests and add more coverage.